### ISSUE TITLE
Unescaped closing bracket ending RegularExpressionLiteral early

### ISSUE THREAD

**Issue by jared-hughes** [lydell/js-tokens#36]
JS-tokens ends the RegularExpressionLiteral when it reaches a `]`, but Firefox and Chrome treat it as if it has been escaped.

```js
> [...jsTokens("/xyz]def/")]
[
  { type: 'RegularExpressionLiteral', value: '/xyz', closed: false },
  { type: 'Punctuator', value: ']' },
  { type: 'IdentifierName', value: 'def' },
  { type: 'Punctuator', value: '/' }
]
> /xyz]def/
/xyz]def/
> [...jsTokens("/xyz]def*/")]
[
  { type: 'RegularExpressionLiteral', value: '/xyz', closed: false },
  { type: 'Punctuator', value: ']' },
  { type: 'IdentifierName', value: 'def' },
  { type: 'Punctuator', value: '*' },
  { type: 'RegularExpressionLiteral', value: '/', closed: false }
]
> /xyz]def*/
/xyz]def*/
```

Encountered in a real file in [fluent.js](https://github.com/projectfluent/fluent.js/blob/e99614eb7fdc0ac6d09ee2cadf8bc05763c51ac1/fluent-bundle/src/resource.ts#L59).

I'm not sure if this is technically a bug with JS-tokens. The regex does not match [the spec](https://tc39.es/ecma262/#sec-literals-regular-expression-literals), so this means Chrome and Firefox go against the spec. However, they will likely not fix this due to backwards compatibility, so it seems more appropriate to update JS-tokens.

**Comment by lydell** [comment]
This might be mentioned in section [B Additional ECMAScript Features for Web Browsers – B.1.2 Regular Expressions Patterns](https://tc39.es/ecma262/#sec-regular-expressions-patterns), but I’d have to read it again to be certain.

Interestingly, `/]/` is valid according to Chrome and Firefox, but `/]/u` is a syntax error. I do remember `/u` enabling stricter parsing rules though.

Either way, my current feeling is that your patch is correct for js-tokens use cases. Will take a deeper look soon.

Out of curiosity, how did you encounter this?

**Comment by jared-hughes** [comment]
I was working with a big file that included fluent.js and hence `const TOKEN_BRACKET_CLOSE = /\s*] */y;`. The processing involved keeping track of bracket nesting depth, and this produced an extraneous close bracket, messing up matching for everything after that.

My current workaround is `.replace("/\\s*] */y", "/\\s*\\] */y")` because this problem only appears once.

**Comment by lydell** [comment]
Found it in the spec:

Annex B extends [Term](https://tc39.es/ecma262/#prod-annexB-Term) with [ExtendedAtom](https://tc39.es/ecma262/#prod-annexB-ExtendedAtom), which contains `[ExtendedPatternCharacter](https://tc39.es/ecma262/#prod-annexB-ExtendedPatternCharacter)` which is basically any character (with a few exceptions, but `]` is not an exception).

**Comment by lydell** [comment]
Thanks for reporting and fixing!

Released in v8.0.1.
